src/query/sql/src/planner/optimizer/rule/rewrite/rule_eager_aggregation.rs

```diff
@@ -0,0 +1,349 @@
+
+        // Divide the aggregate functions into left and right based on the columns in the aggregate function.
+        let aggregations = vec![
+            get_aggregation_functions(&agg_final, &columns_set[0]),
+            get_aggregation_functions(&agg_final, &columns_set[1]),
+        ];
+
+        // dbg!("aggregations = {:?}", &aggregations);
+
+        if aggregations[0].is_empty() && aggregations[1].is_empty() {
+            return Ok(());
+        }
+
+        let mut group_columns_set = vec![ColumnSet::new(), ColumnSet::new()];
+        for item in agg_final.group_columns()?.iter() {
+            for idx in 0..2 {
+                if columns_set[idx].contains(item) {
+                    group_columns_set[idx].insert(*item);
+                }
+            }
+        }
+
+        let mut is_eager = [true, true];
+        for cond in join.left_conditions.iter() {
+            for c in cond.used_columns().iter() {
+                if !group_columns_set[0].contains(c) {
+                    is_eager[0] = false;
+                }
+            }
+        }
+        for cond in join.right_conditions.iter() {
+            for c in cond.used_columns().iter() {
+                if !group_columns_set[1].contains(c) {
+                    is_eager[1] = false;
+                }
+            }
+        }
+
+        // dbg!("group_columns_set = {:?}", &group_columns_set);
+
+        let factory = AggregateFunctionFactory::instance();
+        let d = if aggregations[0].is_empty() { 1 } else { 0 };
+
+        if is_eager[d] && is_eager[d ^ 1] {
+            // TODO(dousir9):
+            // if aggregations[d ^ 1].is_empty() {
+            //     apply double eager on both.
+            // } else {
+            //     apply eager split on both.
+            //     apply eager groupby-count on d.
+            // }
+        } else if is_eager[d] && !is_eager[d ^ 1] {
+            if aggregations[d ^ 1].is_empty() {
+                // Apply eager group by on d
+                let eager_agg_final = agg_final.clone();
+                let eager_agg_partial = agg_partial.clone();
+                for (index, aggregation_function_index, func_name) in aggregations[d].iter() {
+                    if !factory.is_decomposable(func_name) {
+                        continue;
+                    }
+
+                    let aggregation_functions = vec![
+                        &mut agg_partial.aggregate_functions[*index],
+                        &mut agg_final.aggregate_functions[*index],
+                    ];
+                    for aggregate_function in aggregation_functions {
+                        if let ScalarExpr::AggregateFunction(agg) = &mut aggregate_function.scalar {
+                            let new_data_type = match agg.func_name.as_str() {
+                                "count" => Box::new(DataType::Number(NumberDataType::UInt64)),
+                                "sum" => Box::new(DataType::Number(NumberDataType::Int64)),
```

Comment:
Don't use fixed datatype, aggregate function had `return_type`.

Sum is not `Int64` type, could be `f64`, `i64`, `u64`.

Reply:
Okay.

Reply:
Okay.